### ISSUE TITLE
No randomness in the airdrop

### ISSUE THREAD

**Issue by sonya75** [mouse-dev-1/Akutar#3]
There is no randomness in the airdrop tokens if the address[] in airdrop is controlled by the owner. There should be some hash verification to make sure that the address[] isn't shuffled by the owner after the shiftQuantity has been committed. There are any number of ways to implement the check.

**Comment by sonya75** [comment]
The second issue is require(block.number >= blockToUse,"Not enough time has passed to reveal.")
It should be strict inequality. You can't get the blockhash of current block.

**Comment by sonya75** [comment]
reveal() function shouldn't have onlyOwner modifier because otherwise the owner can call commit() any number of the times before using reveal()

**Comment by tannakartikey** [comment]
> The second issue is require(block.number >= blockToUse,"Not enough time has passed to reveal.")
It should be strict inequality. You can't get the blockhash of current block.

@sonya75, the `commit()` function has to be called before `reveal()` which sets `blockToUse`. So when we are calling `reveal()`, we are getting the blockhash of the block which is already created.

**Comment by sonya75** [comment]
No, you are setting blockToUse=block.number+5. That block has not been created when commit() is called.

**Comment by tannakartikey** [comment]
Yes, but blockhash is called only after we reach `blockToUse`. So not trying to get the blockhash of the current block.
>  You can't get the blockhash of current block.

**Comment by sonya75** [comment]
Well, that's only if the reveal function has onlyOwner modifier. Which was my other point, that it shouldn't. Check my third post in this thread.

**Comment by tannakartikey** [comment]
Yes, so if there is a check that we can only call `commit()` once then `onlyOwner` can also be kept and it will be called only once, right?

**Comment by sonya75** [comment]
I don't think that's a good idea either. Because in case you do miss calling the reveal() function within 256 blocks, you won't get the blockhash. The best option would be to use proper check that blockhash value isn't zero, which it will be if 256 blocks pass or if the block hasn't been mined yet.

**Comment by 0xildefonso** [comment]
> I don't think that's a good idea either. Because in case you do miss calling the reveal() function within 256 blocks, you won't get the blockhash. The best option would be to use proper check that blockhash value isn't zero, which it will be if 256 blocks pass or if the block hasn't been mined yet.

guess this is not a huge issue, since deployers can always make sure `reveal` succeeds and then declare this is the canonical contract for akutars. perfect engineering solution? no, but enough to make it work.
